src/PolyType/Abstractions/TypeShapeResolver.cs

```diff
@@ -0,0 +1,184 @@
+    /// </para>
+    /// <para>
+    /// Intended for compatibility with older frameworks where the <see cref="IShapeable{T}"/>
+    /// interface (which uses static abstract interface methods) is not available.
+    /// </para>
+    /// </remarks>
+    public static ITypeShape<T>? ResolveDynamic<
+#if NET
+        [DynamicallyAccessedMembers(ResolveDynamicLinkerRequirements)]
+#endif
+        T>(bool throwIfMissing = false)
+    {
+        ITypeShape<T>? result = ResolveDynamicFactoryCache<T, T>.GetFactory()?.Invoke();
+        if (throwIfMissing && result is null)
+        {
+            Throw();
+            static void Throw() => throw new NotSupportedException($"The type '{typeof(T)}' does not have a generated shape. Ensure that the type is annotated with the 'GenerateShape' attribute.");
+        }
+
+        return result;
+    }
+
+    /// <summary>
+    /// Uses reflection to resolve the source generated <see cref="ITypeShape{T}"/> implementation from another type.
+    /// </summary>
+    /// <typeparam name="T">The type for which to extract the shape.</typeparam>
+    /// <typeparam name="TProvider">The type from which to extract the shape.</typeparam>
+    /// <param name="throwIfMissing">If <see langword="true"/>, throws an exception if no shape is found.</param>
+    /// <returns>An <see cref="ITypeShape{T}"/> instance describing <typeparamref name="T"/>, or <see langword="null"/> if none is found.</returns>
+    /// <exception cref="NotSupportedException"><paramref name="throwIfMissing"/> is <see langword="true"/> and no shape is found.</exception>"
+    /// <remarks>
+    /// <para>
+    /// Uses reflection to look for potential <see cref="IShapeable{T}"/> implementations or
+    /// <see cref="TypeShapeProviderAttribute"/> annotations inserted onto <typeparamref name="T"/>
+    /// by the source generator.
+    /// </para>
+    /// <para>
+    /// Intended for compatibility with older frameworks where the <see cref="IShapeable{T}"/>
+    /// interface (which uses static abstract interface methods) is not available.
+    /// </para>
+    /// </remarks>
+    public static ITypeShape<T>? ResolveDynamic<T,
+#if NET
+        [DynamicallyAccessedMembers(ResolveDynamicLinkerRequirements)]
+#endif
+        TProvider>(bool throwIfMissing = false)
+    {
+        ITypeShape<T>? result = ResolveDynamicFactoryCache<T, TProvider>.GetFactory()?.Invoke();
+        if (throwIfMissing && result is null)
+        {
+            Throw();
+            static void Throw() => throw new NotSupportedException($"The type '{typeof(TProvider)}' does not have a generated shape for '{typeof(T)}'. Ensure that the type is annotated with the 'GenerateShapeFor' attribute.");
+        }
+
+        return result;
+    }
+
+#if NET
+    /// <summary>
+    /// Gets the dynamic access requirements for the <see cref="ResolveDynamic{T}"/> method group.
+    /// </summary>
+    public const DynamicallyAccessedMemberTypes ResolveDynamicLinkerRequirements =
```

Comment:
@AArnott the only drawback of the current `ResolveDynamic` implementation is that it forces all those linker requirements just to enable dynamic `IShapeable<T>` invocation. I've filed an issue about this in the [runtime repo](https://github.com/dotnet/runtime/issues/119440) for potential better alternatives, but I think we could forgo these requirements altogether if we also had the source generator emit `TypeShapeProviderAttribute` in targets that do support `IShapeable<T>`. The obvious drawback of such an approach is that it would automatically root the generated type shapes to their declarations, even if the types are never used with PolyType libraries!